ext/uri/php_uriparser.c

```diff
@@ -0,0 +1,432 @@
+/*
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Máté Kocsis <kocsismate@php.net>                            |
+   +----------------------------------------------------------------------+
+*/
+
+#include "php.h"
+#include "php_uriparser.h"
+#include "php_uri_common.h"
+#include "Zend/zend_smart_str.h"
+#include "Zend/zend_exceptions.h"
+
+static void uriparser_free_uri(void *uri);
+static void throw_invalid_uri_exception(void);
+
+static inline size_t get_text_range_length(const UriTextRangeA *range)
+{
+	return range->afterLast - range->first;
+}
+
+static UriUriA *uriparser_copy_uri(UriUriA *uriparser_uri)
+{
+	ZEND_ASSERT(uriparser_uri != NULL);
+
+	UriUriA *new_uriparser_uri = emalloc(sizeof(UriUriA));
+
+	int result = uriCopyUriA(new_uriparser_uri, uriparser_uri);
+	ZEND_ASSERT(result == URI_SUCCESS && new_uriparser_uri != NULL);
+
+	return new_uriparser_uri;
+}
+
+static void uriparser_normalize_uri(UriUriA *uriparser_uri)
+{
+	ZEND_ASSERT(uriNormalizeSyntaxExA(uriparser_uri, (unsigned int)-1) == URI_SUCCESS);
```

Comment:
In release build, ZEND_ASSERT is transformed into ZEND_ASSUME. However, one some platforms this throws away the expression. For those, the call to `uriNormalizeSyntaxExA` won't happen. You need to call `uriNormalizeSyntaxExA` on a separate line, assign the return value to a variable, and do the assertion on the variable.

Reply:
This _might_ be BTW why your Windows tests fail. (Not checked)

Reply:
This is not the only place where you do this. You have to fix this in all instances.

Reply:
I knew that asserts are sometimes not eliminated even in release builds, when they have a side-effect, so ideally they should be wrapped around an `#if ZEND_DEBUG`.

That's why I thought that it's ok to use them the way I did, although these went through multiple iterations (actual code and assert on separate lines, if condition with the actual code and then ZEND_UNREACHABLE inside the if body), but finally I chose the wrong one. So thanks for the info! And the windows failures seem very likely.